src/main/java/com/google/sps/model/AuthenticatedHttpServlet.java

```diff
@@ -93,4 +123,43 @@ public void doPost(
       throws IOException {
     response.sendError(400, "POST is not supported");
   }
+
+  /**
+   * Get a Google Credential object from the authentication cookies (idToken & accessToken) in the
+   * HTTP Request
+   *
+   * @param request Http Request sent from client
+   * @return Credential object with accessToken. null if tokens not present / are empty
+   * @throws TokenVerificationError if the passed idToken is confirmed to be false / is unverifiable
+   * @throws GeneralSecurityException if an issue occurs with Google's verification service
+   * @throws IOException if an issue occurs with Google's verification service
+   */
+  private Credential getGoogleCredential(HttpServletRequest request)
+      throws TokenVerificationError, GeneralSecurityException, IOException {
+    // If cookies not present, return null (user's session likely expired)
+    Cookie idTokenCookie = ServletUtility.getCookie(request, "idToken");
+    Cookie accessTokenCookie = ServletUtility.getCookie(request, "accessToken");
+    if (idTokenCookie == null || accessTokenCookie == null) {
+      return null;
+    }
+    String idToken = idTokenCookie.getValue();
+    String accessToken = accessTokenCookie.getValue();
+
+    try {
+      if (!authenticationVerifier.verifyUserToken(idToken)) {
+        throw new TokenVerificationError(String.format("idToken (value=%s) is invalid!", idToken));
+      }
+    } catch (GeneralSecurityException | IOException e) {
+      e.printStackTrace();
```

Comment:
You shouldn't have to manually print the stack trace. It's probably not useful outside of active development anyways.

Ideally, whatever framework runs these servlets will log the stack and other data for unhandled exceptions.

Reply:
Interesting! Thanks